Gift Entry: Re-organize GE and FORM Classes

kenlewis-sfdo opened the pull request:
[W-7988305](https://gus.lightning.force.com/lightning/r/ADM_Work__c/a07B0000008URs2IAG/view)

This pull request reduces the number of classes used by the Gift Entry application by refactoring to move methods closer to their usage points and eliminate a number of files in the GE and FORM prefixes.  

Files removed by this PR after relocating their methods:
FORM_BatchHeaderField
FORM_Element
FORM_Layout
FORM_PermissionValidator
FORM_PermissionsValidator_TEST
FORM_RenderWrapper
FORM_Section
FORM_Service
FORM_ServiceGiftEntry
FORM_ServiceGiftEntry_TEST
FORM_Template
GE_DefaultTemplate
GE_DefaultTemplate_TEST
GE_FormRendererService
GE_FormRendererService_TEST
GE_FormServiceController
GE_GiftEntry_UTIL
GE_TemplateBuilderCtrl

Files Added:
GE_Template
GE_Template_TEST

# Critical Changes

# Changes

# Issues Closed

# Community Ideas Delivered

# New Metadata

# Deleted Metadata

voduyemi commented on src/classes/GE_DefaultTemplate.cls:
```diff
@@ -137,13 +137,13 @@ public with sharing class GE_DefaultTemplate {
         populateDevNameMapping(updatedFieldDevNameMap, updatedObjectDevNameMap);
```
We have an unused variable at the top of this class `ELEMENT_TYPE_FIELD`, can we remove it ?

Also the `buildDefaultTemplate` method could use some refactoring because of its size, but I'm not sure if we want to get into it for this WI. What do you think ?

psadie-sfdo replied:
👍  for removing `ELEMENT_TYPE_FIELD`.

What do ya'll think about adding the refactor of specific methods to our refactor backlog quip doc?

kenlewis-sfdo replied:
Agree with removing the unused variable, and with the need to refactor the buildDefaultTemplate method.  

I'll push the update to remove the variable now and I like the idea of adding specific method to our refactor backlog document.  This WI is really specifically geared toward organization/re-location so I want to shy away from refactoring the internals of any methods here.  And I think having these methods queued up in the doc will help us to identify them as good candidates for refactoring when doing design docs for features and changes that will touch these same classes again in the future, which is when I think we should propose/implement the internal refactoring of this and other similar methods.

ptewson-sfdo replied:
I enthusiastically support the above decisions - whenever a potential refactor CAN be isolated to a particular method, that would be a great way to break up the work into small-ish sections and I think clarify the possible regression impacts as well. Nothing new to contribute beyond that. Y'all are smart!

kenlewis-sfdo replied:
👍🏽✅

voduyemi reviewed (changes requested):
Thanks Ken for this clean-up makes everything much easier to read and find! Left a few minor comments around annotations and using the right accessors and some suggestions for potential refactors. Let me know you think!

psadie-sfdo commented on src/classes/FORM_ServiceGiftEntry.cls:
```diff
@@ -423,5 +423,36 @@ public with sharing class FORM_ServiceGiftEntry {
+    *
+    * @return {String} guid: Generated UUID.
+    */
+    public static String generateId() {
```
Could we private this?

```suggestion
    private static String generateId() {
```

kenlewis-sfdo replied:
👍🏽

psadie-sfdo reviewed (changes requested):
Just a couple of things

kenlewis-sfdo commented on src/classes/GE_GiftEntryController.cls:
```diff
@@ -276,5 +276,369 @@ public with sharing class GE_GiftEntryController {
+        }
+    }
+
+    private static final String BOOLEAN_DATA_TYPE = 'BOOLEAN';
```
👍🏽

kenlewis-sfdo commented on src/classes/GE_GiftEntryController.cls:
```diff
@@ -276,5 +276,369 @@ public with sharing class GE_GiftEntryController {
+    * @description Work in Progress to deliver widget behaviors
+    * @return Widget data
+    */
+    public static String convertWidgetDataToObjectJSON(String widgetData) {
```
All good changes!  I'm going to make the change to make it private and @TestVisible.  And I'll add this method to our list of refactor ideas.

kenlewis-sfdo commented on src/classes/GE_GiftEntryController.cls:
```diff
@@ -276,5 +276,369 @@ public with sharing class GE_GiftEntryController {
+     * @param String templateName
+     * @return  A RenderWrapper containing the default template JSON.
+     */
+    public static RenderWrapper getRenderWrapper(String templateName) {
```
Good call!

kenlewis-sfdo commented on src/classes/GE_GiftEntryController.cls:
```diff
@@ -276,5 +276,369 @@ public with sharing class GE_GiftEntryController {
+     *
+     * @return  A RenderWrapper containing the default template JSON.
+     */
+    public static RenderWrapper getDefaultSGERenderWrapper() {
```
👍🏽

kenlewis-sfdo commented on src/classes/GE_GiftEntryController.cls:
```diff
@@ -276,5 +276,369 @@ public with sharing class GE_GiftEntryController {
+
+    private static final String BOOLEAN_DATA_TYPE = 'BOOLEAN';
+
+    @AuraEnabled
```
Yep - good changes!

kenlewis-sfdo commented:
Thanks for all the feedback so far @voduyemi and @psadie-sfdo !  I think I've addressed all of the comments so far, and started to add the mentioned methods to our [refactor doc](https://salesforce.quip.com/NbLTAT7qKRI3#IERACASGJ4r).  Let me know if I missed anything or if you have any further comments!  Thank you!!

ptewson-sfdo commented on src/classes/FORM_ServiceGiftEntry.cls:
```diff
@@ -423,5 +423,36 @@ public with sharing class FORM_ServiceGiftEntry {
+    *
+    * @return {String} guid: Generated UUID.
+    */
+    private static String generateId() {
```
I think we've discussed as a team that this class could maybe have a better name, since many of the methods here (like this one, and the populateColumnHeaders() method that calls it) are supporting the Gift Entry app more than anything form-specific. But I know we are staggering the delivery of all this refactor so not a big deal if we tackle that in a future WI!

kenlewis-sfdo replied:
This is a good point, and illuminates the fact that Gift Entry classes should probably be unified under the GE prefix with the FORM prefix being removed entirely.  I created a [draft PR here](https://github.com/SalesforceFoundation/NPSP/pull/5863) to view what that would look like.  I'll reach out to try to discuss today if we want to incorporate these changes into this branch/PR and bring this work to its logical conclusion within this WI.

ptewson-sfdo commented on src/classes/FORM_ServiceGiftEntry_TEST.cls:
```diff
@@ -42,38 +42,34 @@ public with sharing class FORM_ServiceGiftEntry_TEST {
         FORM_Template template0 = UTIL_UnitTestData_TEST.createSampleTemplate();
         String templateJSON0 = JSON.serialize(template0);
-        fs.storeFormTemplate(null,template0.name, template0.description, template0.version, templateJSON0);
+        FORM_ServiceGiftEntry.storeFormTemplate(null,template0.name, template0.description, template0.version, templateJSON0);
```
I notice we are moving away from instantiating the FORM_Service class in favor of calling static methods here - we frequently talk about wanting to shift the other direction, towards instances of classes that can be mocked and behaviors  can be controlled by the caller. Just want to make sure we are aware of and ok with the switch in the opposite directly here!

ptewson-sfdo reviewed (commented):
Just a few comments/questions  -  this looks really nice!

ptewson-sfdo reviewed (approved):
The updates look good to me!

voduyemi reviewed (approved):
Looks good to me!

psadie-sfdo reviewed (approved):
I think we're good!
